docs/guides/connect-to-state-with-url-hash.md

```diff
@@ -43,6 +43,114 @@ export const useBoundStore = create(
 )
 ```
 
-## CodeSandbox Demo
+### CodeSandbox Demo
 
 https://codesandbox.io/s/zustand-state-with-url-hash-demo-f29b88?file=/src/store/index.ts
+
+## Persist and Connect State with URL Parameters (Example: URL Query Parameters)
+
+This example is for when you want to conditionally connect the state to the URL (here we use URL query parameters) when visiting the website from a populated URL, and always want it synced and consistent with something like localstorage.
+
+If you want the URL params to always populate, the conditional check on `window.location.search.slice(1)` can be removed.
+
+The implementation below will update the URL in place, without refresh, as the relevant states change.
+
+```ts
+import { create } from 'zustand'
+import { persist, StateStorage, createJSONStorage } from 'zustand/middleware'
+
+const persistentStorage: StateStorage = {
+  getItem: (key): string => {
+    // Check URL first
+    if (window.location.search.slice(1)) {
+      const searchParams = new URLSearchParams(window.location.search.slice(1))
+      const storedValue = searchParams.get(key)
+      return JSON.parse(storedValue) // **
+    } else {
+      // Otherwise, we should load from localstorage or alternative storage
+      return JSON.parse(localStorage.getItem(key))
+    }
+  },
+  setItem: (key, newValue): void => {
+    // Check if query params exist at all, can remove check if always want to set URL
+    if (window.location.search.slice(1)) {
+      const searchParams = new URLSearchParams(window.location.search.slice(1))
+      searchParams.set(key, JSON.stringify(newValue)) // **
+      window.history.replaceState(null, null, `?${searchParams.toString()}`)
+    }
+
+    localStorage.setItem(key, JSON.stringify(newValue))
+  },
+  removeItem: (key): void => {
+    const searchParams = new URLSearchParams(window.location.search.slice(1))
+    searchParams.delete(key)
+    window.location.search = searchParams.toString()
+  },
+}
+
+let localAndUrlStore = (set) => ({
+  searchGameDifficulty: {
+    beginner: false,
+    intermediate: true,
+    expert: false,
+  },
+  setSearchGameDifficulty: (difficulty, selection) =>
+    set((state) => ({
+      searchGameDifficulty: {
+        ...state.searchGameDifficulty,
+        [difficulty]: selection,
+      },
+    })),
+
+  searchRatings: {
+    1: false,
+    2: false,
+    3: false,
+    4: true,
+    5: true,
+  },
+  setSearchRatings: (rating, selection) =>
+    set((state) => ({
+      searchRatings: { ...state.searchRatings, [rating]: selection },
+    })),
```

Comment:
Could you prepare a simpler example? It took me a fair amount of time to understand the domain of the problem. Something obvious would be more suitable, because the example needs to depict the state in URL, not the game search problem.